File: MetaGeta/MetaGeta.DataStore/DataStore/DataStoreManager.vb
Imports MetaGeta.DataStore.Database

Public Class DataStoreManager
    Implements INotifyPropertyChanged
    Implements IDataStoreOwner

    Private Shared ReadOnly log As log4net.ILog = log4net.LogManager.GetLogger(Reflection.MethodBase.GetCurrentMethod().DeclaringType)

    Private ReadOnly m_DataMapper As IDataMapper
    Private ReadOnly m_DataStores As New ObservableCollection(Of MGDataStore)
    Private ReadOnly m_JobQueue As JobQueue

    Public Sub New(ByVal designMode As Boolean)
        log.InfoFormat("DataStoreManager ctor")

        If Not designMode Then
            m_DataMapper = New DataMapper("metageta.db3")
        Else
            m_DataMapper = New MockDataMapper()
        End If

        m_JobQueue = New JobQueue(designMode:=designMode)

        m_DataMapper.Initialize()
        m_DataStores.AddRange(m_DataMapper.GetDataStores(Me))
        OnDataStoresChanged()
    End Sub

    Public Sub WaitForQueueToEmpty()
        m_JobQueue.WaitForQueueToEmpty()
    End Sub

    Public Sub Shutdown()
        m_JobQueue.Dispose()
        For Each ds In m_DataStores
            ds.Close()
        Next
        m_DataMapper.Close()
    End Sub

    Public Function NewDataStore(ByVal name As String, ByVal template As IDataStoreTemplate) As MGDataStore
        Dim data As New MGDataStore(Me, m_DataMapper)
        Using (data.SuspendUpdates())
            data.Template = template
            data.Name = name
        End Using
        For Each pluginTypeName In template.GetPluginTypeNames()
            data.AddNewPlugin(pluginTypeName)
        Next
        m_DataMapper.WriteNewDataStore(data)
        m_DataStores.Add(data)
        OnDataStoresChanged()
        Return data
    End Function

    Public Sub DeleteDataStore(ByVal dataStore As MGDataStore) Implements IDataStoreOwner.DeleteDataStore
        dataStore.Close()
        m_DataMapper.RemoveDataStore(dataStore)
        DataStores.Remove(dataStore)
    End Sub

    Public Sub EnqueueAction(ByVal action As String, ByVal dataStore As MGDataStore, ByVal file As MGFile) Implements IDataStoreOwner.EnqueueAction
        m_JobQueue.EnqueueAction(action, dataStore, file)
    End Sub

    Public ReadOnly Property JobQueue() As JobQueue
        Get
            Return m_JobQueue
        End Get
    End Property

    Public ReadOnly Property DataStores() As ObservableCollection(Of MGDataStore)
        Get
            Return m_DataStores
        End Get
    End Property

    Private Sub OnDataStoresChanged()
        OnPropertyChanged("DataStores")
    End Sub
    Private Sub OnPropertyChanged(ByVal name As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(name))
    End Sub
    Public Event PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs) Implements System.ComponentModel.INotifyPropertyChanged.PropertyChanged
End Class

File: MetaGeta/MetaGeta.DataStore/DataStore/DataStoreManager.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Reflection;
using log4net;
using MetaGeta.DataStore.Database;
using MetaGeta.Utilities;

#endregion

namespace MetaGeta.DataStore {
    public class DataStoreManager : INotifyPropertyChanged, IDataStoreOwner {
        private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
        private readonly IDataMapper m_DataMapper;
        private readonly ObservableCollection<MGDataStore> m_DataStores = new ObservableCollection<MGDataStore>();

        private readonly JobQueue m_JobQueue;

        public DataStoreManager(bool designMode) {
            log.InfoFormat("DataStoreManager ctor");

            if (!designMode)
                m_DataMapper = new DataMapper("metageta.db3");
            else
                m_DataMapper = new MockDataMapper();

            m_JobQueue = new JobQueue(designMode);

            m_DataMapper.Initialize();
            m_DataStores.AddRange(m_DataMapper.GetDataStores(this));
            OnDataStoresChanged();
        }

        public JobQueue JobQueue {
            get { return m_JobQueue; }
        }

        public ObservableCollection<MGDataStore> DataStores {
            get { return m_DataStores; }
        }

        #region IDataStoreOwner Members

        public void DeleteDataStore(MGDataStore dataStore) {
            dataStore.Close();
            m_DataMapper.RemoveDataStore(dataStore);
            DataStores.Remove(dataStore);
        }

        public void EnqueueAction(IAction action, MGDataStore dataStore, MGFile file) {
            m_JobQueue.EnqueueAction(action, dataStore, file);
        }

        #endregion

        #region INotifyPropertyChanged Members

        public event PropertyChangedEventHandler PropertyChanged;

        #endregion

        public void WaitForQueueToEmpty() {
            m_JobQueue.WaitForQueueToEmpty();
        }

        public void Shutdown() {
            m_JobQueue.Dispose();
            foreach (MGDataStore ds in m_DataStores)
                ds.Close();
            m_DataMapper.Close();
        }

        public MGDataStore NewDataStore(string name, IDataStoreTemplate template) {
            var data = new MGDataStore(this, m_DataMapper);
            using ((data.SuspendUpdates())) {
                data.Template = template;
                data.Name = name;
            }
            foreach (string pluginTypeName in template.GetPluginTypeNames())
                data.AddNewPlugin(pluginTypeName);
            m_DataMapper.WriteNewDataStore(data);
            m_DataStores.Add(data);
            OnDataStoresChanged();
            return data;
        }

        private void OnDataStoresChanged() {
            OnPropertyChanged("DataStores");
        }

        private void OnPropertyChanged(string name) {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(name));
        }
    }
}